components/service/glean/src/main/java/mozilla/components/service/glean/StringListMetricType.kt

```diff
@@ -0,0 +1,105 @@
+ *
+ * The string list API exposes the [add] and [set] methods, which take care of validating the input
+ * data and making sure that limits are enforced.
+ */
+data class StringListMetricType(
+    override val disabled: Boolean,
+    override val category: String,
+    override val lifetime: Lifetime,
+    override val name: String,
+    override val sendInPings: List<String>
+) : CommonMetricData {
+
+    override val defaultStorageDestinations: List<String> = listOf("metrics")
+
+    private val logger = Logger("glean/StringListMetricType")
+
+    companion object {
+        // Maximum length of any passed value string, in UTF8 byte sequence length.
+        const val MAX_STRING_LENGTH = 50
+    }
+
+    /**
+     * Appends a string value to one or more string list metric stores.  If the string exceeds the
+     * maximum string length, defined as [MAX_STRING_LENGTH], it will be truncated.
+     *
+     * If adding the string to the lists would exceed the maximum value defined as
+     * [StringListsStorageEngine.MAX_LIST_LENGTH_VALUE], then the storage engine will drop the new
+     * value and it will not be added to the list.
+     *
+     * @param value This is a user defined string value. The maximum length of
+     * this string is [MAX_STRING_LENGTH].
+     */
+    fun add(value: String) {
+        // TODO report errors through other special metrics handled by the SDK. See bug 1499761.
+
+        if (!shouldRecord(logger)) {
+            return
+        }
+
+        val truncatedValue = value.let {
+            if (it.length > MAX_STRING_LENGTH) {
+                logger.warn("$category.$name - string too " +
+                        "long ${it.length} > $MAX_STRING_LENGTH")
+                return@let it.substring(0, MAX_STRING_LENGTH)
+            }
+            it
+        }
+
+        Dispatchers.API.launch {
+            // Delegate storing the string to the storage engine.
+            StringListsStorageEngine.add(
+                    this@StringListMetricType,
+                    value = listOf(truncatedValue)
+            )
+        }
+    }
+
+    /**
+     * Sets a List<String> to one or more metric stores.
+     *
+     * @param value This is a user defined string list. The maximum length of each string in the
+     * list is defined by [MAX_STRING_LENGTH], while the maximum length of the list itself is
+     * defined by [StringListsStorageEngine.MAX_LIST_LENGTH_VALUE].  If a longer list is passed
+     * into this function, then the additional values will be dropped from the list and the list,
+     * up to the [StringListsStorageEngine.MAX_LIST_LENGTH_VALUE], will still be recorded.
+     */
+    fun set(value: List<String>) {
+        // TODO report errors through other special metrics handled by the SDK. See bug 1499761.
+
+        if (!shouldRecord(logger)) {
+            return
+        }
+
+        val stringList = value.map let@{
```

Comment:
Yes, I get an error about undefined @let at the "return@let" if I don't have the let@ label here.  This was the IDE suggested remedy to that error.

Reply:
Ah, cool. I think you can try to change `return@let` to `return` only and drop the `let@` part.

Reply:
Managed to reduce this to `value.map { it.take(MAX_STRING_LENGTH) }`

Reply:
Cool :) The only problem with that, though, is that it's not logging when entries are longer than the maximum string length :(

Reply:
That's true.  I tried this with just `return` dropping the `let@` and `@let` parts but it didn't like it.  How about something more like this:
```
        val stringList = value.map { stringValue ->
            stringValue.takeUnless { it.length > MAX_STRING_LENGTH }
                ?: stringValue.take(MAX_STRING_LENGTH)
                    .also {
                        logger.warn("$category.$name - string too long " +
                                    "${stringValue.length} > $MAX_STRING_LENGTH")
                    }
        }
```

Reply:
FYI, I had to name the `it` to `stringValue` to prevent shadowing of `it`

Reply:
I have the feeling this should work. I can't try that though, as I don't have the environment here on my laptop :(

```
val stringList = value.map {  
  if (it.length > MAX_STRING_LENGTH) {
    logger.warn("$category.$name - string too long ${stringValue.length} > $MAX_STRING_LENGTH")
  }
  it.substring(0, MAX_STRING_LENGTH)
}
```

Reply:
It does, with one minor change to use `.take` instead of `.substring` to prevent the index out of bounds exception.  I'm not sure why I keep wanting to make this one difficult ;)